Enabling haServices causes operator crash

Enabling haServices causes the operator container to panic and crash. We believe this may be due to the new deployment names that are that are created when enabling haServices for clouddriver. 

Operator version 1.0.2 in basic deployment.

```
        haServices:                                                                                                                                                                                                          
          clouddriver:
            disableClouddriverRoDeck: false
            enabled: true
          echo:
            enabled: false
```

```
E0728 17:15:39.111067       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/opt/spinnaker-operator/build/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/opt/spinnaker-operator/build/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/opt/spinnaker-operator/build/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/panic.go:679
/usr/local/go/src/runtime/panic.go:199
/usr/local/go/src/runtime/signal_unix.go:394
/opt/spinnaker-operator/build/pkg/util/find.go:88
/opt/spinnaker-operator/build/pkg/deploy/spindeploy/transformer/accounts.go:92
/opt/spinnaker-operator/build/pkg/deploy/spindeploy/transformer/accounts.go:50
/opt/spinnaker-operator/build/pkg/deploy/spindeploy/spindeploy.go:93
/opt/spinnaker-operator/build/pkg/controller/spinnakerservice/spinnakerservice_controller.go:132
/opt/spinnaker-operator/build/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:216
/opt/spinnaker-operator/build/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:192
/opt/spinnaker-operator/build/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:171
/opt/spinnaker-operator/build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152
/opt/spinnaker-operator/build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153
/opt/spinnaker-operator/build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:1357
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x268 pc=0x13f72d6]

goroutine 787 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/opt/spinnaker-operator/build/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x105
panic(0x1722360, 0x28b3d80)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/armory/spinnaker-operator/pkg/util.GetContainerInDeployment(0x0, 0x194cf3a, 0xb, 0x0)
	/opt/spinnaker-operator/build/pkg/util/find.go:88 +0x26
github.com/armory/spinnaker-operator/pkg/deploy/spindeploy/transformer.addSpringProfile(0x0, 0x194cf3a, 0xb, 0x19492cc, 0x8, 0x8010104, 0x0)
	/opt/spinnaker-operator/build/pkg/deploy/spindeploy/transformer/accounts.go:92 +0x5a
github.com/armory/spinnaker-operator/pkg/deploy/spindeploy/transformer.(*accountsTransformer).TransformManifests(0xc000b0bec0, 0x1cb0e40, 0xc0005302d0, 0xc0003a5260, 0xc0005c2278, 0x0, 0x0)
	/opt/spinnaker-operator/build/pkg/deploy/spindeploy/transformer/accounts.go:50 +0x110
github.com/armory/spinnaker-operator/pkg/deploy/spindeploy.(*Deployer).Deploy(0xc000117a40, 0x1cb0e40, 0xc0005302d0, 0x1cef720, 0xc000136280, 0xc0003a5260, 0x1e, 0x1961628, 0x1a)
	/opt/spinnaker-operator/build/pkg/deploy/spindeploy/spindeploy.go:93 +0x68d
github.com/armory/spinnaker-operator/pkg/controller/spinnakerservice.(*ReconcileSpinnakerService).Reconcile(0xc0002595e0, 0xc0005d36b0, 0xc, 0xc0005d3650, 0x9, 0xc00006ec00, 0x0, 0x0, 0x0)
	/opt/spinnaker-operator/build/pkg/controller/spinnakerservice/spinnakerservice_controller.go:132 +0x70f
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00049ae60, 0x178e4c0, 0xc00082a460, 0x0)
	/opt/spinnaker-operator/build/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:216 +0x162
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00049ae60, 0x0)
	/opt/spinnaker-operator/build/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:192 +0xcb
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc00049ae60)
	/opt/spinnaker-operator/build/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:171 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00039abd0)
	/opt/spinnaker-operator/build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00039abd0, 0x3b9aca00, 0x0, 0x1, 0xc00004c7e0)
	/opt/spinnaker-operator/build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc00039abd0, 0x3b9aca00, 0xc00004c7e0)
	/opt/spinnaker-operator/build/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/opt/spinnaker-operator/build/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157 +0x32e
```

When ha is successfully enabled we typically see the following deployment names: 

spin-clouddriver-caching
spin-clouddriver-ro
spin-clouddriver-ro-deck
spin-clouddriver-rw

--- a/pkg/deploy/spindeploy/transformer/accounts.go
+++ b/pkg/deploy/spindeploy/transformer/accounts.go
@@ -46,9 +46,12 @@ func (a *accountsTransformer) TransformManifests(ctx context.Context, gen *gener
 	}
 
 	// Enable "accounts" Spring profile for each potential service
-	for _, s := range accounts.GetAllServicesWithAccounts() {
-		if err := addSpringProfile(gen.Config[s].Deployment, s, accounts.SpringProfile); err != nil {
-			return err
+	svcs := accounts.GetAllServicesWithAccounts()
+	for k := range gen.Config {
+		if a.isServiceWithAccount(k, svcs) {
+			if err := addSpringProfile(gen.Config[k].Deployment, k, accounts.SpringProfile); err != nil {
+				return err
+			}
 		}
 	}
 
@@ -64,6 +67,15 @@ func (a *accountsTransformer) TransformManifests(ctx context.Context, gen *gener
 	}
 	a.log.Info(fmt.Sprintf("found %d accounts to deploy", len(crdAccs)))
 	return updateServiceSettings(ctx, crdAccs, gen)
+}
+
+func (a *accountsTransformer) isServiceWithAccount(serviceKey string, accountServices []string) bool {
+	for _, s := range accountServices {
+		if util.IsServiceLike(serviceKey, s) {
+			return true
+		}
+	}
+	return false
 }
 
 func updateServiceSettings(ctx context.Context, crdAccounts []account.Account, gen *generated.SpinnakerGeneratedConfig) error {

--- a/pkg/deploy/spindeploy/transformer/accounts_test.go
+++ b/pkg/deploy/spindeploy/transformer/accounts_test.go
@@ -49,6 +49,40 @@ func TestAddSpringProfile(t *testing.T) {
 	}
 }
 
+func TestAddSpringProfileHA(t *testing.T) {
+
+	d := &appsv1.Deployment{
+		Spec: appsv1.DeploymentSpec{
+			Template: v1.PodTemplateSpec{
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name: "clouddriver-ro",
+							Env: []v1.EnvVar{
+								{
+									Name:  "SOME_ENV",
+									Value: "test",
+								},
+								{
+									Name:  "SPRING_PROFILES_ACTIVE",
+									Value: "local,test",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	err := addSpringProfile(d, "clouddriver-ro", "accounts")
+	assert.Nil(t, err)
+	c := util.GetContainerInDeployment(d, "clouddriver-ro")
+	if assert.NotNil(t, c) {
+		assert.Equal(t, 2, len(c.Env))
+		assert.Equal(t, "local,test,accounts", c.Env[1].Value)
+	}
+}
+
 func TestAddSpringProfileExisting(t *testing.T) {
 	d := &appsv1.Deployment{
 		Spec: appsv1.DeploymentSpec{

--- a/pkg/util/find.go
+++ b/pkg/util/find.go
@@ -14,6 +14,7 @@ import (
 	clientcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
 	"k8s.io/client-go/rest"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
 )
 
 var errSecretNotFound = errors.New("secret not found")
@@ -92,6 +93,10 @@ func GetContainerInDeployment(dep *v12.Deployment, containerName string) *v1.Con
 		}
 	}
 	return nil
+}
+
+func IsServiceLike(svc1, svc2 string) bool {
+	return svc1 == svc2 || strings.HasPrefix(svc1, svc2+"-")
 }
 
 func UpdateSecret(secret *v1.Secret, svc string, settings map[string]interface{}, profileName string) error {

--- a/pkg/util/find_test.go
+++ b/pkg/util/find_test.go
@@ -110,3 +110,43 @@ spec:
 	assert.Equal(t, "spin-echo-files-287979322", s.ObjectMeta.Name)
 	assert.Contains(t, s.Data, "echo.yml")
 }
+
+func TestServiceLike(t *testing.T) {
+	cases := []struct {
+		name string
+		svc1 string
+		svc2 string
+		like bool
+	}{
+		{
+			"different services",
+			"gate",
+			"clouddriver",
+			false,
+		},
+		{
+			"same services",
+			"gate",
+			"gate",
+			true,
+		},
+		{
+			"specialized service",
+			"echo-scheduler",
+			"echo",
+			true,
+		},
+		{
+			"almost specialized service",
+			"echos-cheduler", // Typo on purpose
+			"echo",
+			false,
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			assert.Equal(t, c.like, IsServiceLike(c.svc1, c.svc2))
+		})
+	}
+}

